src/node/stage/comm.rs

```diff
@@ -6,136 +6,447 @@
+            let conn = self.endpoint.connect_to(recipient).await?;
+            let conn = Arc::new(conn);
+            let _ = self
+                .node_conns
+                .lock()
+                .await
+                .insert(*recipient, Arc::clone(&conn));
+
+            conn
+        };
+
+        conn.send_uni(msg).await?;
+
+        Ok(())
     }
+}
 
-    pub async fn send_direct_message(
-        &self,
-        src_id: &FullId,
-        recipient: &SocketAddr,
-        variant: Variant,
-    ) -> Result<()> {
-        let message = Message::single_src(src_id, DstLocation::Direct, variant, None, None)?;
-        self.send_message_to_target(recipient, message.to_bytes())
+// Helper to track the sending of a single message to potentially multiple recipients.
+struct SendState {
+    recipients: Vec<Recipient>,
+    remaining: usize,
+}
+
+struct Recipient {
+    addr: SocketAddr,
+    sending: bool,
+    attempt: u8,
+}
+
+impl SendState {
+    fn new(recipients: &[SocketAddr], delivery_group_size: usize) -> Self {
+        Self {
+            recipients: recipients
+                .iter()
+                .map(|addr| Recipient {
+                    addr: *addr,
+                    sending: false,
+                    attempt: 0,
+                })
+                .collect(),
+            remaining: delivery_group_size,
+        }
+    }
+
+    // Returns the next recipient to send to.
+    fn next(&mut self) -> Option<SocketAddr> {
+        let active = self
+            .recipients
+            .iter()
+            .filter(|recipient| recipient.sending)
+            .count();
+
+        if active >= self.remaining {
+            return None;
+        }
+
+        let recipient = self
+            .recipients
+            .iter_mut()
+            .filter(|recipient| !recipient.sending && recipient.attempt < RESEND_MAX_ATTEMPTS)
+            .min_by_key(|recipient| recipient.attempt)?;
+
+        recipient.attempt += 1;
+        recipient.sending = true;
+
+        Some(recipient.addr)
+    }
+
+    // Marks the recipient as failed.
+    fn failure(&mut self, addr: &SocketAddr) {
+        if let Some(recipient) = self
+            .recipients
+            .iter_mut()
+            .find(|recipient| recipient.addr == *addr)
+        {
+            recipient.sending = false;
+        }
+    }
+
+    // Marks the recipient as successful.
+    fn success(&mut self, addr: &SocketAddr) {
+        if let Some(index) = self
+            .recipients
+            .iter()
+            .position(|recipient| recipient.addr == *addr)
+        {
+            let _ = self.recipients.swap_remove(index);
+            self.remaining -= 1;
+        }
+    }
+
+    // Consumes the state and returns the list of recipients that failed all attempts (if any).
+    fn finish(self) -> Vec<SocketAddr> {
+        self.recipients
+            .into_iter()
+            .filter(|recipient| !recipient.sending && recipient.attempt >= RESEND_MAX_ATTEMPTS)
```

Comment:
if a recipient is still sending, it will not be counted as `failed` ?
but after `finish` being called, the state is supposed to be consumed and that `recipient`'s result won't be handled?

Reply:
Yeah, you are not really supposed to call `finish` while there are still sends in progress, but the api doesn't prevent that either.

Reply:
Maybe we can count sending as failed too. Not sure. Doesn't really matter in practice, because we never call `finish` while there are still pending sends.